vllm/core/scheduler.py

```diff
@@ -395,6 +395,38 @@ def __init__(
         # for processing and deallocation by the free_finished_seq_groups()
         self._async_stopped: List[SequenceGroup] = []
 
+        # For chunked prefill, we allocate a set of "prefill slots" that
+        # each represent a sequence group that can be partially prefilled.
+        # Having multiple partial prefills in flight allows us to minimize TTFT
+        # and avoid decode starvation in cases where a single sequence group
+        # with a very large prompt blocks the queue for too many iterations.
+        self.num_prefill_slots = scheduler_config.num_prefill_slots
+        self.prefill_slots_running = 0
+        self.big_prefill_requests = 0
+        # Requests with more than (4% max context length) tokens to prefill
+        # are "big".
+        # The number of big prefill requests is limited so that smaller
+        # requests may jump the queue in front of them and get to the decode
+        # phase faster.
+        self.big_prefill_threshold = scheduler_config.max_model_len // 25
+        self.max_big_requests = 1  # TODO: something
+
+        # Dict cache with the chunk sizes to hand out to each sequence depending
+        # on how many prefill slots are used. This is slightly faster than
+        # running an integer division every time a prefill is scheduled.
+        # This splits the budget evenly among all prefill slots.
+        # We use a defaultdict here to handle the case where we prefill many
+        # more requests than prefill slots. This is normal when requests have
+        # very small prompts to prefill, and we want to give them each the same
+        # budget.
+        self.prefill_chunk_sizes = defaultdict(
+            lambda: scheduler_config.max_num_batched_tokens // self.
+            num_prefill_slots)
+        self.prefill_chunk_sizes[0] = scheduler_config.max_num_batched_tokens
+        for i in range(1, self.num_prefill_slots):
+            self.prefill_chunk_sizes[i] = \
+                scheduler_config.max_num_batched_tokens // i
```

Comment:
The reasoning is in the docstring- it's a defaultdict so that when there are potentially many prefills running at the same time we cap the minimum per-slot budget at `budget // self.num_prefill_slots`

We can use a list instead though and try to ensure that we never access past the end